### PR TITLE
Parallel Builds

### PR DESCRIPTION
**Pull request by krauthaufen**
Added support for parallel builds by using an environment-variable %fake-parallel-jobs%

Note that this of course only works when target-functions can be executed concurrently.
The feature can be enabled by passing for example "fake-parallel-jobs=4" to FAKE.exe

Since parallel build cause problems in TraceHelpers the current "stack" of opened tags is now thread-local.

### CODE REVIEW

File: src/app/FakeLib/TargetHelper.fs [hunk 1]
```diff
@@ -322,6 +324,41 @@ let listTargets() =
...
+
+    traverse [d] 0
+
+
+    let result = 
+        levels |> Seq.map (fun (KeyValue(l,p)) -> (l,p))
+                |> Seq.groupBy snd
+                |> Seq.sortBy (fun (l,_) -> -l)
+                |> Seq.map snd
+                |> Seq.map (fun v -> v |> Seq.map fst |> Set.ofSeq |> Seq.map getTarget |> Seq.toArray)
```
Comment (vasily-kirichenko) [inline src/app/FakeLib/TargetHelper.fs]: Maybe `Seq.distinct` is more clear here?

File: src/app/FakeLib/TargetHelper.fs [hunk 2]
```diff
@@ -345,13 +382,72 @@ let run targetName =
                     traceEndTarget target.Name                
         with
         | exn -> targetError targetName exn
+
+    let parallelJobs = environVarOrDefault "fake-parallel-jobs" "1" |> Int32.Parse
```
Comment (vasily-kirichenko) [inline src/app/FakeLib/TargetHelper.fs]: `Int32.Parse` -> `int`?

File: src/app/FakeLib/TargetHelper.fs [hunk 3]
```diff
@@ -345,13 +382,72 @@ let run targetName =
...
       
     let watch = new System.Diagnostics.Stopwatch()
     watch.Start()        
     try
         tracefn "Building project with version: %s" buildVersion
-        PrintDependencyGraph false targetName
-        runTarget targetName
+
+        if parallelJobs > 1 then
```
Comment (vasily-kirichenko) [inline src/app/FakeLib/TargetHelper.fs]: The body of this `if` is very long and hard to read. What about extracting a function or two?

File: src/app/FakeLib/TargetHelper.fs [hunk 4]
```diff
@@ -345,13 +382,72 @@ let run targetName =
...
+                                finally
+                                    finished.Signal() |> ignore
+                            | _ ->
+                                System.Threading.Thread.Sleep(100)
+                with :? System.OperationCanceledException ->
+                    ()
+
+            let workers =
+                Array.init parallelJobs (fun i -> 
+                    let thread = new System.Threading.Thread(System.Threading.ThreadStart(worker), IsBackground = true)
```
Comment (vasily-kirichenko) [inline src/app/FakeLib/TargetHelper.fs]: As you use unordered `ConcurrentBag` for targets, I think it's better to use the thread pool instead of dedicated threads. Just create an `async` for each `Target` and start all of them in parallel (`Async.Parallel >> Async.RunSynchronously`). To set degree of parallelism you could use a `SemaphoreSlim`. Another advantage is that cancellation is handled automatically. As an alternative, you could use raw TPL (`seq of targets -> AsParallel -> WithDegreeOfParallelism -> ToArray`).

File: src/app/FakeLib/TargetHelper.fs [hunk 5]
```diff
@@ -322,6 +324,41 @@ let listTargets() =
 // Instead of the target can be used the list dependencies graph parameter.
 let doesTargetMeansListTargets target = target = "--listTargets"  || target = "-lt"
 
+
+let determineBuildOrder (d : TargetTemplate<unit>) =
+    let levels = Dictionary<string, int>()
```
Comment (vasily-kirichenko) [inline src/app/FakeLib/TargetHelper.fs]: Maybe use `Seq.fold` and a `Map` instead of mutable dic and imperative loop?

File: src/app/FakeLib/TargetHelper.fs [hunk 6]
```diff
@@ -345,13 +382,72 @@ let run targetName =
...
-        runTarget targetName
+
+        if parallelJobs > 1 then
+            // multi threaded build
+
+            tracefn "Running parallel build with %d workers" parallelJobs
+
+            let queue = System.Collections.Concurrent.ConcurrentBag()
+            let finished = new System.Threading.CountdownEvent(0)
+            let cancel = new System.Threading.CancellationTokenSource()
```
Comment (vasily-kirichenko) [inline src/app/FakeLib/TargetHelper.fs]: let -> use

**Comment by krauthaufen** [comment]
Hi there, I reworked the entire thing a little according to your review.
Sorry for the long code-segments but I'm used to the F# Outlining extension collapsing those functions/etc.

**Comment by krauthaufen** [comment]
Hi vasily, I finally managed to change everything according to your review.
I decided to use the TPL directly since I'm not quite sure about the semantics of Async.Parallel

However all the other points should be adressed now. Please let me know if you have any further suggestions.

File: src/app/FakeLib/TargetHelper.fs [hunk 7]
```diff
@@ -322,36 +324,114 @@ let listTargets() =
...
+            let watch = new System.Diagnostics.Stopwatch()
+            watch.Start()
+            target.Function()
+            addExecutedTarget target.Name watch.Elapsed
+            traceEndTarget target.Name     
+    with exn ->
+        targetError target.Name exn
+
+/// Runs the given array of targets in parallel using count tasks
+let runTargetsParallel (count : int) (targets : Target[]) =
```
Comment (vasily-kirichenko) [inline src/app/FakeLib/TargetHelper.fs]: I think we can do better :) What about
```fsharp
let runTargetsParallel (count : int) (targets : Target[]) =
    targets.AsParallel()
        .WithDegreeOfParallelism(count)
        .Select(runSingleTarget)
        .ToArray()
    |> ignore
```
or
```fsharp
let runTargetsParallel (count : int) (targets : Target[]) =
    Parallel.ForEach(targets, ParallelOptions(MaxDegreeOfParallelism = count), runSingleTarget)
```
?

Comment (krauthaufen) [inline src/app/FakeLib/TargetHelper.fs]: Yes this is obviously more concise but I didn't want to use "C# style Linq stuff" here.
I liked your first fragment better for some reason :)

File: src/app/FakeLib/TargetHelper.fs [hunk 8]
```diff
@@ -322,36 +324,114 @@ let listTargets() =
 // Instead of the target can be used the list dependencies graph parameter.
 let doesTargetMeansListTargets target = target = "--listTargets"  || target = "-lt"
```
Comment (vasily-kirichenko) [inline src/app/FakeLib/TargetHelper.fs]: means -> mean :)

File: src/app/FakeLib/TargetHelper.fs [hunk 9]
```diff
@@ -322,36 +324,114 @@ let listTargets() =
...
 
+/// Determines a parallel build order for the given set of targets
+let determineBuildOrder (d : TargetTemplate<unit>) =
```
Comment (vasily-kirichenko) [inline src/app/FakeLib/TargetHelper.fs]: This function is quite complex. I think it's a good idea to write some unit tests for it.

File: src/app/FakeLib/TargetHelper.fs (hunk 7 above)
Comment (vasily-kirichenko) [inline src/app/FakeLib/TargetHelper.fs]: It turns out that the both LINQ solutions uses the thread from which they are executed as one of the worker threads. This is what it's like if it's run from main thread (on a 4 cores box):

image

I think it's totally ok for this case, but you should understand that not all targets will be executed in Thread Pool.

File: src/app/FakeLib/TargetHelper.fs (hunk 8 above)
Comment (krauthaufen) [inline src/app/FakeLib/TargetHelper.fs]: not my function

File: src/app/FakeLib/TargetHelper.fs (hunk 9 above)
Comment (krauthaufen) [inline src/app/FakeLib/TargetHelper.fs]: unit test for traversing an acyclic graph? really? :-D

File: src/app/FakeLib/TargetHelper.fs (hunk 7 above)
Comment (krauthaufen) [inline src/app/FakeLib/TargetHelper.fs]: I'm familiar with .NETs TPL and thats why i said everything will be running on its own Task (not thread) which of course means that also the main thread can be used by the scheduler.
But conceptually (when considering Tasks as abstract threads) everything runs on its own Task...

File: src/app/FakeLib/TargetHelper.fs (hunk 9 above)
Comment (forki) [inline src/app/FakeLib/TargetHelper.fs]: please add unit tests. Sure, they don't prove anything about the function, but experience shows they work very well in practice. Don't forget we have more than 100 contributors (and much more users) from very different fields (a lot don't have any computer science background). Having a set of tests which checks basic edge cases will make it much easier for future contributors to maintain this function. 

If you still think unit tests are useless here, then please look at [FsCheck](https://fsharp.github.io/FsCheck/) and try to prove some important properties. We already have a FsCheck project at https://github.com/fsharp/FAKE/tree/master/src/test/FsCheck.Fake.

Comment (krauthaufen) [inline src/app/FakeLib/TargetHelper.fs]: Okay I'm convinced that Tests could help refactoring this function appropriately.
I'm just not sure if the test code will be mich easier to unterstand than the function itself...

Comment (forki) [inline src/app/FakeLib/TargetHelper.fs]: it's not about understanding one side alone. It's more like [double-entry bookkeeping](http://en.wikipedia.org/wiki/Double-entry_bookkeeping_system). Both sides play together.

Comment (krauthaufen) [inline src/app/FakeLib/TargetHelper.fs]: The problem I'm having is that generators for arbitrary depdendency graphs do not compose with the global mutable state (namely TargetDict).
Nevertheless it would be cool to have arbitrary dependency graphs for testing...

The only solution I came up with is having a different dependency graph and augmenting determineBuildOrder with functions for extracting dependencies..

Any thoughts on that?

Comment (krauthaufen) [inline src/app/FakeLib/TargetHelper.fs]: Hi there,
I implemented some very basic tests for the parallel build order checking if it maintains dependencies correctly.
However I did generate the graphs by hand and cleared the TargetDict manually before running the test.
Please let me know if you have any further suggestions.

**Comment by krauthaufen** [comment]
Hi there,
I implemented some very basic tests for the parallel build order checking if it maintains dependencies correctly.
However I did generate the graphs by hand and cleared the TargetDict manually before running the test.
Please let me know if you have any further suggestions.

File: src/test/FsCheck.Fake/TestParallelBuildOrder.fs [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    "a" ==> "dep" |> ignore
+    "b" ==> "dep" |> ignore
+    "c" ==> "dep" |> ignore
+
+    let order = determineBuildOrder "dep"
+
+    validateBuildOrder order "dep"
+
+    match order with
+        | [f; [|d|]] ->
```
Comment (vasily-kirichenko) [inline src/test/FsCheck.Fake/TestParallelBuildOrder.fs]: Maybe it's worth to introduce an active pattern `Target`, which'd enable syntax like

```fsharp
match order with
| [Target "dep"; [| Target "a"; Target "b"; Target "c" |] -> ...
| _ -> failwith ...
```

File: src/test/FsCheck.Fake/TestParallelBuildOrder.fs [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    "a" ==> "b" ==> "d" |> ignore
+    "a" ==> "c" ==> "d" |> ignore
+
+    let order = determineBuildOrder "d"
+    validateBuildOrder order "d"
+
+    match order with
+        | [[|a|];[|b;c|];[|d|]] ->
+            let bcNamesCorrect =
+                if b.Name = "b" then c.Name = "c"
```
Comment (vasily-kirichenko) [inline src/test/FsCheck.Fake/TestParallelBuildOrder.fs]: Same as above.

File: src/test/FsCheck.Fake/TestParallelBuildOrder.fs (hunk 10 above)
Comment (krauthaufen) [inline src/test/FsCheck.Fake/TestParallelBuildOrder.fs]: Hi Vasily, I like the idea but the problem with active patterns is that the set of targets on each level is basically unordered which means that 
``
[| Target "a"; Target "b" |]
``

would need to be wrapped into a set-pattern like:
``
TargetSet ["a"; "b"; "c"]
``
where all target-names are (for instance) ordered alphabetically...
I'm going to work on that in the evening...

Let me know if you have any problems with that solution...

**Comment by krauthaufen** [comment]
Hi folks, the tests are now using active patterns (as described above)

**Comment by vasily-kirichenko** [comment]
Looks good for me, thanks! :)

**Comment by forki** [comment]
Awesome! :heart:  

So the only thing that's left is a small tutorial.
Could you please come up with something similar to 
https://github.com/fsharp/FAKE/blob/master/help/specifictargets.md

**Comment by krauthaufen** [comment]
added just a little documentation and a list of issues:
https://github.com/krauthaufen/FAKE/blob/master/help/parallelbuild.md

**Comment by forki** [comment]
very cool. I'm getting excited.

alk

Only one small thing: Can we use `parallel-jobs` instead of `fake-parallel-jobs`?

**Comment by krauthaufen** [comment]
done

**Comment by forki** [comment]
released in 3.21

let's see how it goes.

parallel

thanks for all the effort.

**Comment by Thorium** [comment]
How do I configure dependency tree to take advantage of parallel run?
Is this ok:

    Task_1
      ==> Task_2_A
      ==> Task_3
    
    Task_1
      ==> Task_2_B
      ==> Task_3

Will it run Task_2_A and Task_2_B in parallel if I have e.g. parallel-jobs=2 ?

**Comment by krauthaufen** [comment]
Yes it should indeed find a build-plan where those tasks are executed in parallel...
The graph is basically traversed in level-order which causes some tasks to be started not at the earliest possible time but should give a a reasonable approximation for the build order...

The algorithm starts with the final target and assigns all targets a maximal distance to that...
In the end everything having the same level will be executed in parallel...

Hope that helps
